### DISCUSSION TITLE
Tool Change Procedure

### DISCUSSION THREAD

**Discussion by Miniclubbin** [vlachoudis/bCNC#1686] [General]
Tool change is found under the 'Probe' tab and further 'Tool' tab. I use a button mounted to my wasteboard, but I believe any probe will work.

### Configuration:
**Fast Probe Feed:** worthless, changing this doesn't affect tool change
**Probe Feed:** Probe feedrate from defined probe location
**Probe Command:** choose your preference

**Policy:** Manual Tool Change (WCS) - this changes your WCS Z0 based on tool length difference
**Pause:** choose your pause preference

**Change:** MCS coordinates of your tool change location- mine is top Z and middle of my work area. You can set this manually or jog to the exact location and hit 'get' to record it automatically
**Probe:** MCS coordinates of your tool probe location, including a safe Z position for your longest tool. Jog your machine to the location and Z height and hit 'get' to record it automatically
**Distance:** Amount of probe travel in negative Z, based on probe location defined above
**Calibration:** Automatic value for location of probe contact below WCS Z0

### Steps:

1. Install first tool and set WCS 0 positions, _including Z height_
2. Before running program (or after, but before tool change), hit '**calibrate**' on top right of probe toolbar or bottom calibration button
	a. Tool moves to MCS change position as defined above
3. Run first job
4. On tool change Gcode command or finished job, **ensure tool is in safe WCS position above material** and hit '**Change**' on the top right of probe toolbar
	a. Tool moves to MCS change position as defined above
5. Physically change your tool
6. Hit '**pause**' in the control tab to initiate probe sequence
	a. Tool will move to MCS XYZ probe location and probe at '**probe feed**' rate defined above
	b. WCS Z0 will automatically be reset to account for difference in tool length
	c. Tool will return to last WCS location, wherever that was before you hit 'change'
	d. Tool will continue job without prompts if your Gcode has tool change commands. 
	e. If you want it to wait for you to hit play, select a different 'Pause' option above
7. Run next job

![image](https://user-images.githubusercontent.com/75689479/155258380-95fcc8b4-c8c9-4df5-92cc-f764a243237a.png)

**Comment by MARIOBASZ** [comment]
I consider a better option to use TLO tool change instead of WCS: if you use different work areas (G54 - G55 and following) TLO is consistent with it

**Reply by Miniclubbin** [reply]
I didn't notice the TLO option doing anything, how would I use it?

**Reply by JipeB** [reply]
I share the opinion of MarioBasz, TLO is the only procedure implemented in GRBL compatible with the different work plans.
To avoid any erratic operation of bCNC with the TLO procedure I suggest adding these 2 lines in the CNC.py file 

lines.append("G49")	# reset TLO  The current TLO must be canceled before mounting a new tool.
lines.append("%wait")        Waiting for cancellation confirmation from GRBL

        #----------------------------------------------------------------------
	# code to change manually tool
	#----------------------------------------------------------------------
	def toolChange(self, tool=None):
		if tool is not None:
			# Force a change
			self.tool = tool
			self._lastTool = None

		# check if it is the same tool
		if self.tool is None or self.tool == self._lastTool: return []

		# create the necessary code
		lines = []
		lines.append("$g")	# remember state and populate variables
		lines.append("m5")	# stop spindle
		lines.append("%wait")
	>>	lines.append("G49")	# reset TLO
	>>	lines.append("%wait")
		lines.append("%_x,_y,_z = wx,wy,wz")	# remember position
		lines.append("g53 g0 z[toolchangez]")
		lines.append("g53 g0 x[toolchangex] y[toolchangey]")
		lines.append("%wait")

For the implementation of the procedure:
[https://www.youtube.com/watch?v=4QMUFPKhjys](url)
It's in French, but with the pictures...
JP
